assets/js/modules/analytics-4/datastore/report.js

```diff
@@ -152,6 +156,104 @@ const baseSelectors = {
 
 		return reports[ stringifyObject( options ) ];
 	},
+
+	/**
+	 * Determines whether the Analytics 4 module is still gathering data.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return {boolean|undefined} Returns `true` if gathering data, otherwise `false`. Returns `undefined` while resolving.
+	 */
+	isGatheringData: createRegistrySelector( ( select ) => () => {
+		const { startDate, endDate } = select( CORE_USER ).getDateRangeDates( {
+			offsetDays: DATE_RANGE_OFFSET,
+		} );
+
+		const args = {
+			dimensions: [ 'date' ],
+			metrics: [ { name: 'totalUsers' } ],
+			startDate,
+			endDate,
+		};
+
+		const url = select( CORE_SITE ).getCurrentEntityURL();
+		if ( url ) {
+			args.url = url;
+		}
+
+		// Disable reason: select needs to be called here or it will never run.
+		// eslint-disable-next-line @wordpress/no-unused-vars-before-return
+		const report = select( MODULES_ANALYTICS_4 ).getReport( args );
+		const hasResolvedReport = select(
+			MODULES_ANALYTICS_4
+		).hasFinishedResolution( 'getReport', [ args ] );
+
+		if ( ! hasResolvedReport ) {
+			return undefined;
+		}
+
+		if ( ! isPlainObject( report ) ) {
+			return false;
+		}
+
+		if ( ! Array.isArray( report?.rows ) || report?.rows?.length === 0 ) {
+			return true;
+		}
+
+		return false;
+	} ),
+
+	/**
+	 * Determines whether Analytics 4 has zero data or not.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return {boolean|undefined} Returns FALSE if not gathering data and the report is not zero, otherwise TRUE. If the request is still being resolved, returns undefined.
+	 */
+	hasZeroData: createRegistrySelector( ( select ) => () => {
+		const { startDate, endDate } = select( CORE_USER ).getDateRangeDates( {
+			offsetDays: DATE_RANGE_OFFSET,
+		} );
+
+		const args = {
+			dimensions: [ 'date' ],
+			metrics: [ { name: 'totalUsers' } ],
+			startDate,
+			endDate,
+		};
+
+		const url = select( CORE_SITE ).getCurrentEntityURL();
+		if ( url ) {
+			args.url = url;
+		}
```

Comment:
Let's shuffle this a little bit. These functionality isn't needed during gathering state check, right? Then, let's move it after the gathering data status check.

Reply:
The same logic applies as in my reply to your comment on `isGatheringData()`, except in this case we don't have clear changes in the pipeline to `hasZeroData()`.

I'd suggest either leaving this as-is, or applying it both here and to the UA `hasZeroData()`. What do you think?

Reply:
Yes, let's apply for both.

Reply:
I've applied both of the suggested changes to `hasZeroData()`.